File: backend/src/controllers/rideController.ts
import { Response } from 'express';
import { prisma } from '../config/db';
import { AuthRequest } from '../middleware/authMiddleware';
import { calculateFare, getAllFareEstimates } from '../services/fareService';
import { RideType } from '@prisma/client';
import { io } from '../index';

export const estimateFare = async (req: AuthRequest, res: Response) => {
  try {
    const { pickupLat, pickupLng, destLat, destLng } = req.body;

    if (!pickupLat || !pickupLng || !destLat || !destLng) {
      return res.status(400).json({ error: 'Pickup and Destination coordinates required' });
    }

    // Distance calculation using Haversine formula (in km)
    const distanceKm = calculateHaversineDistance(
      parseFloat(pickupLat),
      parseFloat(pickupLng),
      parseFloat(destLat),
      parseFloat(destLng)
    );

    // Travel time estimate based on average 30 km/h speed
    const estimatedTimeMin = Math.max(5, Math.round((distanceKm / 30) * 60));

    const estimates = await getAllFareEstimates(distanceKm, estimatedTimeMin);

    return res.json({
      distanceKm: Math.round(distanceKm * 10) / 10,
      estimatedTimeMin,
      estimates
    });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to calculate fare estimate', details: error.message });
  }
};

export const createRide = async (req: AuthRequest, res: Response) => {
  try {
    if (!req.user) return res.status(401).json({ error: 'Authentication required' });

    const {
      pickupLat,
      pickupLng,
      pickupAddress,
      destLat,
      destLng,
      destAddress,
      rideType,
      paymentMethod
    } = req.body;

    if (!pickupLat || !pickupLng || !pickupAddress || !destLat || !destLng || !destAddress || !rideType) {
      return res.status(400).json({ error: 'Missing required ride parameters' });
    }

    const distanceKm = calculateHaversineDistance(
      parseFloat(pickupLat),
      parseFloat(pickupLng),
      parseFloat(destLat),
      parseFloat(destLng)
    );

    const estimatedTimeMin = Math.max(5, Math.round((distanceKm / 30) * 60));
    const selectedType = (rideType as RideType) || 'SEDAN';

    const fare = await calculateFare(selectedType, distanceKm, estimatedTimeMin);

    const rideNumber = `URM-${Math.floor(10000 + Math.random() * 90000)}`;

    const ride = await prisma.ride.create({
      data: {
        rideNumber,
        passengerId: req.user.id,
        pickupLat: parseFloat(pickupLat),
        pickupLng: parseFloat(pickupLng),
        pickupAddress,
        destLat: parseFloat(destLat),
        destLng: parseFloat(destLng),
        destAddress,
        rideType: selectedType,
        status: 'SEARCHING_DRIVER',
        distanceKm: Math.round(distanceKm * 10) / 10,
        estimatedTimeMin,
        baseFare: fare.baseFare,
        perKmFare: fare.perKmFare,
        estimatedFare: fare.estimatedFare,
        paymentMethod: paymentMethod || 'CASH',
        paymentStatus: 'PENDING'
      },
      include: {
        passenger: {
          select: { id: true, name: true, phone: true, email: true }
        }
      }
    });

    // Create payment entry
    await prisma.payment.create({
      data: {
        rideId: ride.id,
        amount: fare.estimatedFare,
        paymentMethod: paymentMethod || 'CASH',
        status: 'PENDING'
      }
    });

    // Find nearby online drivers matching vehicle type
    const nearbyDrivers = await prisma.driver.findMany({
      where: {
        isOnline: true,
        vehicle: {
          type: selectedType
        }
      },
      include: {
        user: { select: { name: true, phone: true } },
        vehicle: true
      }
    });

    // Broadcast WebSocket notification to drivers
    io.emit('ride:new_request', {
      ride,
      availableDriversCount: nearbyDrivers.length
    });

    return res.status(201).json({
      message: 'Ride request created, searching for nearby drivers...',
      ride
    });
  } catch (error: any) {
    console.error('Error creating ride:', error);
    return res.status(500).json({ error: 'Failed to create ride request', details: error.message });
  }
};

export const acceptRide = async (req: AuthRequest, res: Response) => {
  try {
    if (!req.user || req.user.role !== 'DRIVER') {
      return res.status(403).json({ error: 'Only drivers can accept rides' });
    }

    const { id } = req.params;
    const driverId = req.user.driverId;

    if (!driverId) {
      return res.status(400).json({ error: 'Driver profile not found' });
    }

    const ride = await prisma.ride.findUnique({ where: { id } });
    if (!ride) return res.status(404).json({ error: 'Ride not found' });

    if (
      ride.status !== 'REQUESTED' &&
      ride.status !== 'SEARCHING_DRIVER' &&
      !(ride.status === 'DRIVER_ASSIGNED' && ride.driverId === driverId)
    ) {
      return res.status(400).json({ error: `Ride cannot be accepted in state ${ride.status}` });
    }

    const updatedRide = await prisma.ride.update({
      where: { id },
      data: {
        driverId,
        status: 'DRIVER_ASSIGNED',
        assignedAt: new Date()
      },
      include: {
        passenger: { select: { id: true, name: true, phone: true, email: true } },
        driver: {
          include: {
            user: { select: { name: true, phone: true } },
            vehicle: true
          }
        }
      }
    });

    // Broadcast WebSocket updates
    io.to(`ride:${id}`).emit('ride:status:change', {
      rideId: id,
      status: 'DRIVER_ASSIGNED',
      ride: updatedRide
    });

    io.emit('ride:status:change', {
      rideId: id,
      status: 'DRIVER_ASSIGNED',
      ride: updatedRide
    });

    return res.json({ message: 'Ride accepted successfully', ride: updatedRide });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to accept ride', details: error.message });
  }
};

export const markArrived = async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;
    const ride = await prisma.ride.update({
      where: { id },
      data: {
        status: 'DRIVER_ARRIVED',
        arrivedAt: new Date()
      },
      include: {
        passenger: { select: { id: true, name: true, phone: true } },
        driver: { include: { user: { select: { name: true, phone: true } }, vehicle: true } }
      }
    });

    io.to(`ride:${id}`).emit('ride:status:change', { rideId: id, status: 'DRIVER_ARRIVED', ride });
    io.emit('ride:status:change', { rideId: id, status: 'DRIVER_ARRIVED', ride });

    return res.json({ message: 'Driver marked as arrived', ride });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to update status' });
  }
};

export const startRide = async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;
    const ride = await prisma.ride.update({
      where: { id },
      data: {
        status: 'RIDE_STARTED',
        startedAt: new Date()
      },
      include: {
        passenger: { select: { id: true, name: true, phone: true } },
        driver: { include: { user: { select: { name: true, phone: true } }, vehicle: true } }
      }
    });

    io.to(`ride:${id}`).emit('ride:status:change', { rideId: id, status: 'RIDE_STARTED', ride });
    io.emit('ride:status:change', { rideId: id, status: 'RIDE_STARTED', ride });

    return res.json({ message: 'Ride started', ride });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to start ride' });
  }
};

export const completeRide = async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;
    const ride = await prisma.ride.findUnique({ where: { id } });
    if (!ride) return res.status(404).json({ error: 'Ride not found' });

    const finalFare = ride.estimatedFare;

    const updatedRide = await prisma.ride.update({
      where: { id },
      data: {
        status: 'COMPLETED',
        finalFare,
        paymentStatus: 'SUCCESS',
        completedAt: new Date()
      },
      include: {
        passenger: { select: { id: true, name: true, phone: true } },
        driver: { include: { user: { select: { name: true, phone: true } }, vehicle: true } }
      }
    });

    // Update payment record
    await prisma.payment.updateMany({
      where: { rideId: id },
      data: { status: 'SUCCESS', amount: finalFare, transactionId: `TXN-${Date.now()}` }
    });

    // Update driver earnings and total rides
    if (ride.driverId) {
      await prisma.driver.update({
        where: { id: ride.driverId },
        data: {
          totalRides: { increment: 1 },
          totalEarnings: { increment: finalFare }
        }
      });
    }

    io.to(`ride:${id}`).emit('ride:status:change', { rideId: id, status: 'COMPLETED', ride: updatedRide });
    io.emit('ride:status:change', { rideId: id, status: 'COMPLETED', ride: updatedRide });

    return res.json({ message: 'Ride completed successfully', ride: updatedRide });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to complete ride' });
  }
};

export const cancelRide = async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;
    const { reason } = req.body;
    const isDriver = req.user?.role === 'DRIVER';

    const status = isDriver ? 'CANCELLED_BY_DRIVER' : 'CANCELLED_BY_RIDER';

    const ride = await prisma.ride.update({
      where: { id },
      data: {
        status,
        cancelledAt: new Date(),
        cancelReason: reason || (isDriver ? 'Driver cancelled request' : 'Passenger cancelled request')
      },
      include: {
        passenger: { select: { id: true, name: true } },
        driver: { select: { id: true } }
      }
    });

    io.to(`ride:${id}`).emit('ride:status:change', { rideId: id, status, ride });
    io.emit('ride:status:change', { rideId: id, status, ride });

    return res.json({ message: 'Ride cancelled', ride });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to cancel ride' });
  }
};

export const getRide = async (req: AuthRequest, res: Response) => {
  try {
    const { id } = req.params;
    const ride = await prisma.ride.findUnique({
      where: { id },
      include: {
        passenger: { select: { id: true, name: true, phone: true, email: true } },
        driver: {
          include: {
            user: { select: { name: true, phone: true } },
            vehicle: true
          }
        },
        payment: true,
        rating: true
      }
    });

    if (!ride) return res.status(404).json({ error: 'Ride not found' });
    return res.json({ ride });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to fetch ride' });
  }
};

export const listRides = async (req: AuthRequest, res: Response) => {
  try {
    if (!req.user) return res.status(401).json({ error: 'Authentication required' });

    let whereClause: any = {};

    if (req.user.role === 'PASSENGER') {
      whereClause.passengerId = req.user.id;
    } else if (req.user.role === 'DRIVER' && req.user.driverId) {
      whereClause.driverId = req.user.driverId;
    }

    const rides = await prisma.ride.findMany({
      where: whereClause,
      orderBy: { requestedAt: 'desc' },
      include: {
        passenger: { select: { name: true, phone: true } },
        driver: {
          include: {
            user: { select: { name: true } },
            vehicle: true
          }
        },
        payment: true,
        rating: true
      }
    });

    return res.json({ rides });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to list rides' });
  }
};

// Helper Haversine formula for distance in KM
function calculateHaversineDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371; // Earth's radius in km
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

File: backend-spring/src/main/java/com/urbanride/controller/RideController.java
package com.urbanride.controller;

import com.urbanride.dto.CancelRideRequest;
import com.urbanride.dto.CreateRideRequest;
import com.urbanride.dto.EstimateRequest;
import com.urbanride.entity.Ride;
import com.urbanride.repository.RideRepository;

import com.urbanride.security.JwtUtil;
import com.urbanride.service.FareService;
import com.urbanride.service.RideService;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/rides")
public class RideController {

    @Autowired
    private RideService rideService;

    @Autowired
    private FareService fareService;

    @Autowired
    private RideRepository rideRepository;

    @Autowired
    private JwtUtil jwtUtil;

    private String extractToken(HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            return authHeader.substring(7);
        }
        return null;
    }

    @PostMapping("/estimate")
    public ResponseEntity<?> estimateFare(@RequestBody EstimateRequest request) {
        try {
            if (request.getPickupLat() == null || request.getPickupLng() == null ||
                request.getDestLat() == null || request.getDestLng() == null) {
                return ResponseEntity.badRequest().body(Map.of("error", "Pickup and Destination coordinates required"));
            }

            double distanceKm = rideService.calculateHaversineDistance(
                    request.getPickupLat(), request.getPickupLng(),
                    request.getDestLat(), request.getDestLng()
            );
            int estimatedTimeMin = Math.max(5, (int) Math.round((distanceKm / 30.0) * 60));
            var estimates = fareService.getAllFareEstimates(distanceKm, estimatedTimeMin);

            Map<String, Object> response = new HashMap<>();
            response.put("distanceKm", Math.round(distanceKm * 10.0) / 10.0);
            response.put("estimatedTimeMin", estimatedTimeMin);
            response.put("estimates", estimates);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping
    public ResponseEntity<?> createRide(@RequestBody CreateRideRequest request, HttpServletRequest httpRequest) {
        try {
            String token = extractToken(httpRequest);
            if (token == null) return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
            String passengerId = jwtUtil.extractUserId(token);

            Ride ride = rideService.createRide(passengerId, request);

            return ResponseEntity.status(201).body(Map.of(
                    "message", "Ride request created, searching for nearby drivers...",
                    "ride", ride
            ));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to create ride request", "details", e.getMessage()));
        }
    }

    @PostMapping("/{id}/accept")
    public ResponseEntity<?> acceptRide(@PathVariable String id, HttpServletRequest httpRequest) {
        try {
            String token = extractToken(httpRequest);
            if (token == null) return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));

            String driverUserId = jwtUtil.extractUserId(token);
            String driverProfileId = jwtUtil.extractDriverId(token);

            Ride ride = rideService.acceptRide(id, driverUserId, driverProfileId);
            return ResponseEntity.ok(Map.of("message", "Ride accepted successfully", "ride", ride));
        } catch (Exception e) {
            return ResponseEntity.status(400).body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/{id}/arrive")
    public ResponseEntity<?> markArrived(@PathVariable String id) {
        try {
            Ride ride = rideService.markArrived(id);
            return ResponseEntity.ok(Map.of("message", "Driver marked as arrived", "ride", ride));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/{id}/start")
    public ResponseEntity<?> startRide(@PathVariable String id) {
        try {
            Ride ride = rideService.startRide(id);
            return ResponseEntity.ok(Map.of("message", "Ride started", "ride", ride));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/{id}/complete")
    public ResponseEntity<?> completeRide(@PathVariable String id) {
        try {
            Ride ride = rideService.completeRide(id);
            return ResponseEntity.ok(Map.of("message", "Ride completed successfully", "ride", ride));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/{id}/cancel")
    public ResponseEntity<?> cancelRide(@PathVariable String id, @RequestBody(required = false) CancelRideRequest request, HttpServletRequest httpRequest) {
        try {
            String token = extractToken(httpRequest);
            String role = token != null ? jwtUtil.extractRole(token) : "PASSENGER";
            boolean isDriver = "DRIVER".equalsIgnoreCase(role);

            String reason = request != null ? request.getReason() : null;
            Ride ride = rideService.cancelRide(id, reason, isDriver);
            return ResponseEntity.ok(Map.of("message", "Ride cancelled", "ride", ride));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", e.getMessage()));
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getRide(@PathVariable String id) {
        Ride ride = rideRepository.findById(id).orElse(null);
        if (ride == null) return ResponseEntity.status(404).body(Map.of("error", "Ride not found"));
        return ResponseEntity.ok(Map.of("ride", ride));
    }

    @GetMapping
    public ResponseEntity<?> listRides(HttpServletRequest httpRequest) {
        try {
            String token = extractToken(httpRequest);
            if (token == null) return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));

            String userId = jwtUtil.extractUserId(token);
            String role = jwtUtil.extractRole(token);
            String driverId = jwtUtil.extractDriverId(token);

            List<Ride> rides = rideService.listRides(userId, role, driverId);
            return ResponseEntity.ok(Map.of("rides", rides));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", e.getMessage()));
        }
    }
}
